src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,151 @@
+            requests.append(
+                RequestedQuota(
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
+
+
+class WritesLimiter:
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[UseCaseKey, RedisSlidingWindowRateLimiter] = {}
+
+    def _get_rate_limiter(self, use_case_id: UseCaseKey) -> RedisSlidingWindowRateLimiter:
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisSlidingWindowRateLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
+
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[Tuple[KeyResult, FetchType, FetchTypeExt]]]:
+        """
+        Takes a KeyCollection and applies DB write limits as configured via sentry.options.
+
+        Returns a tuple of:
+
+        1. `RateLimitState` that needs to be passed to `apply_write_limits` in order
+          to commit quotas upon successful DB write.
+
+        2. A key collection containing all unmapped keys that passed through the
+          rate limiter.
+
+        3. All unmapped keys that did not pass through the rate limiter.
+        """
+
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
+        dropped_key_results = []
+
+        for org_id, request, grant in zip(org_ids, requests, grants):
+            allowed_strings = granted_key_collection[org_id]
+            if len(allowed_strings) > grant.granted:
+                allowed_strings = set(allowed_strings)
+
+                while len(allowed_strings) > grant.granted:
+                    dropped_key_results.append(
+                        (
+                            KeyResult(org_id=org_id, string=allowed_strings.pop(), id=None),
+                            FetchType.RATE_LIMITED,
+                            FetchTypeExt(
+                                is_global=any(
+                                    quota.prefix_override is not None
+                                    for quota in grant.reached_quotas
+                                ),
+                            ),
+                        )
+                    )
+
+            granted_key_collection[org_id] = allowed_strings
+
+        state = use_case_id, requests, grants, timestamp
+        return state, KeyCollection(granted_key_collection), dropped_key_results
```

Comment:
Why do we need to return the use case id and the requests ?

Reply:
this `state` is an "opaque" value that the indexer shouldn't poke at and pass to `apply_write_limits`. I thought this would be the best way to ensure we read and write with the same usecase.